tracing/src/main/java/org/axonframework/extensions/tracing/TracingQueryGateway.java

```diff
@@ -84,13 +85,13 @@ public static Builder builder() {
 
     @Override
     public <R, Q> CompletableFuture<R> query(String queryName, Q query, ResponseType<R> responseType) {
-        Span parentSpan = tracer.activeSpan();
-        Span span = tracer.buildSpan("send_" + SpanUtils.messageName(query.getClass(), queryName)).start();
-        try (Scope ignored = tracer.activateSpan(span)) {
-            return delegate.query(queryName, query, responseType).whenComplete((r, e) -> {
-                span.finish();
-            });
-        }
+        QueryMessage qry = new GenericQueryMessage(query, queryName, responseType);
```

Comment:
There's a minute caveat, which is pretty important though.
Users can provide either their `queryPayload` as the `query` parameter, or a `QueryMessage` directly.
The idea behind this is to support the option to provide custom `MetaData` when sending the query on the gateway.

Thus, we need to check whether the given `query` is of type `Message`, as otherwise we'd wrap it twice.
The way to achieve this is to use the `GenericMessage#asMessage(Object)` method.

You can check the `DefaultQueryGateway#query` method implementation for an example of how this is done.

Reply:
Good one, made the adjustments but it required a bump on AxonFramework itself since `GenericMessage#asMessage(Object)` is available from 4.3 onwards.